File: zstacktestagent/zstacktestagent/plugins/host.py
'''

@author: Frank
'''
import zstacktestagent.testagent  as testagent
import zstacklib.utils.shell      as shell
import zstacklib.utils.jsonobject as jsonobject
import zstacklib.utils.http       as http
import zstacklib.utils.linux      as linux
import zstacklib.utils.log        as log

logger = log.get_logger(__name__)

class CreateVlanDeviceCmd(testagent.AgentCommand):
    def __init__(self):
        super(CreateVlanDeviceCmd, self).__init__()
        self.ethname = None
        self.vlan = None
        self.ip = None
        self.netmask = None

class SetDeviceIpCmd(testagent.AgentCommand):
    def __init__(self):
        super(SetDeviceIpCmd, self).__init__()
        self.ethname = None
        self.ip = None
        self.netmask = None

class FlushDeviceIpCmd(testagent.AgentCommand):
    def __init__(self):
        super(FlushDeviceIpCmd, self).__init__()
        self.ethname = None

class DeleteVlanDeviceCmd(testagent.AgentCommand):
    def __init__(self):
        super(DeleteVlanDeviceCmd, self).__init__()
        self.vlan_ethname = None

class DeleteBridgeCmd(testagent.AgentCommand):
    def __init__(self):
        super(DeleteVlanDeviceCmd, self).__init__()
        self.bridge_name = None

class HostShellCmd(testagent.AgentCommand):
    def __init__(self):
        super(HostShellCmd, self).__init__()
        self.command = None

class HostShellRsp(testagent.AgentResponse):
    def __init__(self):
        super(HostShellRsp, self).__init__()
        self.return_code = None
        self.stdout = None
        self.stderr = None
        self.command = None
        
CREATE_VLAN_DEVICE_PATH = '/host/createvlandevice'
DELETE_VLAN_DEVICE_PATH = '/host/deletevlandevice'
DELETE_BRIDGE_PATH = '/host/deletebridgedevice'
SET_DEVICE_IP_PATH = '/host/setdeviceip'
FLUSH_DEVICE_IP_PATH = '/host/flushdeviceip'
HOST_SHELL_CMD_PATH = '/host/shellcmd'
ECHO_PATH = '/host/echo'

class HostAgent(testagent.TestAgent):
    def start(self):
        testagent.TestAgentServer.http_server.register_sync_uri(CREATE_VLAN_DEVICE_PATH, self.create_vlan_device)
        testagent.TestAgentServer.http_server.register_sync_uri(DELETE_VLAN_DEVICE_PATH, self.delete_vlan_device)
        testagent.TestAgentServer.http_server.register_sync_uri(DELETE_BRIDGE_PATH, self.delete_bridge)
        testagent.TestAgentServer.http_server.register_sync_uri(SET_DEVICE_IP_PATH, self.set_device_ip)
        testagent.TestAgentServer.http_server.register_sync_uri(FLUSH_DEVICE_IP_PATH, self.flush_device_ip)
        testagent.TestAgentServer.http_server.register_sync_uri(HOST_SHELL_CMD_PATH, self.host_shell_cmd)
        testagent.TestAgentServer.http_server.register_sync_uri(ECHO_PATH, self.echo)
    
    def stop(self):
        pass

    @testagent.replyerror
    def echo(self):
        logger.debug('echo ping')
        #rsp = testagent.AgentResponse()
        return 
    
    @testagent.replyerror
    def create_vlan_device(self, req):
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        linux.create_vlan_eth(cmd.ethname, cmd.vlan, cmd.ip_, cmd.netmask_)
        rsp = testagent.AgentResponse()
        logger.debug('create vlan device: %s' % cmd.vlan)
        return jsonobject.dumps(rsp)
    
    @testagent.replyerror
    def set_device_ip(self, req):
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        linux.set_device_ip(cmd.ethname, cmd.ip, cmd.netmask)
        rsp = testagent.AgentResponse()
        return jsonobject.dumps(rsp)

    @testagent.replyerror
    def flush_device_ip(self, req):
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        linux.flush_device_ip(cmd.ethname)
        rsp = testagent.AgentResponse()
        return jsonobject.dumps(rsp)

    @testagent.replyerror
    def delete_vlan_device(self, req):
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        linux.delete_vlan_eth(cmd.vlan_ethname)
        rsp = testagent.AgentResponse()
        logger.debug('delete vlan device: %s' % cmd.vlan)
        return jsonobject.dumps(rsp)
    
    @testagent.replyerror
    def delete_bridge(self, req):
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        linux.delete_bridge(cmd.bridge_name)
        rsp = testagent.AgentResponse()
        return jsonobject.dumps(rsp)
    
    @testagent.replyerror
    def host_shell_cmd(self, req):
        cmd = jsonobject.loads(req[http.REQUEST_BODY])
        scmd = shell.ShellCmd(cmd.command)
        scmd(False)
        rsp = HostShellRsp()
        rsp.return_code = scmd.return_code
        rsp.stderr = scmd.stderr
        rsp.stdout = scmd.stdout
        rsp.command = cmd.command
        logger.debug('execute %s, the stdout: %s, the stderr: %s' % (cmd.command, scmd.stdout, scmd.stderr))
        return jsonobject.dumps(rsp)

File: zstacktestagent/testagent/src/testagent/plugin/host.go
package plugin

import (
	"testagent/server"
	"testagent/utils"
	"fmt"
	"strings"
)


type HostDeleteBridgeCmd struct {
	BridgeName string `json:"bridge_name"`
}

type HostShellCmd struct {
	Command string `json:"command"`
}

type HostShellRsp struct {
	ReturnCode int `json:"return_code"`
	Stderr string `json:"stderr"`
	Stdout string `json:"stdout"`
	Command string `json:"command"`
	Success bool `json:"success"`
}

type HostCreateVlanDeviceCmd struct {
	Ethname string `json:"ethname"`
	Vlan int `json:"vlan"`
	Ip_ string `json:"ip_"`
	Netmask_ string `json:"netmask_"`
}

type HostDeleteVlanDeviceCmd struct {
	VlanEthname string `json:"vlan_ethname"`
}

type HostSetDeviceIpCmd struct {
	Ethname string `json:"ethname"`
	Ip string `json:"ip"`
	Netmask string `json:"netmask"`
}

type HostFlushDeviceIpCmd struct {
	Ethname string `json:"ethname"`
}

const (
	CREATE_VLAN_DEVICE_PATH = "/host/createvlandevice"
	DELETE_VLAN_DEVICE_PATH = "/host/deletevlandevice"
	DELETE_BRIDGE_PATH = "/host/deletebridgedevice"
	SET_DEVICE_IP_PATH = "/host/setdeviceip"
	FLUSH_DEVICE_IP_PATH = "/host/flushdeviceip"
	HOST_SHELL_CMD_PATH = "/host/shellcmd"
	HOST_VYOS_CMD_PATH = "/host/vyoscmd"
//	HOST_ECHO_PATH = "/host/echo"
)


func hostEchoHandler(ctx *server.CommandContext) interface{} {
	return nil
}

func DeleteVlanEth(VlanDevName string) {
	if isNetworkDeviceExisting(VlanDevName) {
		b := utils.NewBash()
		b.Command = fmt.Sprintf("ip link set dev %s down", VlanDevName)
		b.Run()

		b.Command = fmt.Sprintf("vconfig rem %s", VlanDevName)
		b.Run()
	}
}

func hostCreateVlanDeviceHandler(ctx *server.CommandContext) interface{} {
	b := utils.NewBash()
	cmd := &HostCreateVlanDeviceCmd{}
	ctx.GetCommand(cmd)

	if isNetworkDeviceExisting(cmd.Ethname) {
		VlanDevName := fmt.Sprintf("%s.%s", cmd.Ethname, cmd.Vlan)
		if !isNetworkDeviceExisting(VlanDevName) {
			b.Command = fmt.Sprintf("vconfig add %s %s", cmd.Ethname, strconv.Itoa(cmd.Vlan))
			b.Run()
			if cmd.Ip_ != "" {
				b.Command = fmt.Sprintf("ifconfig %s %s netmask %s", cmd.Ethname, VlanDevName, cmd.Netmask_)
				b.Run()
			}
		} else {
			if GetDeviceIp(VlanDevName) != cmd.Ip_ {
				DeleteVlanEth(VlanDevName)
				b.Command = fmt.Sprintf("vconfig add %s %s", cmd.Ethname, strconv.Itoa(cmd.Vlan))
				b.Run()
				b.Command = fmt.Sprintf("ifconfig %s %s netmask %s", VlanDevName, cmd.Ip_, cmd.Netmask_)
				b.Run()
			}
		}
	}
		
	return nil
}


func hostDeleteVlanDeviceHandler(ctx *server.CommandContext) interface{} {
	cmd := &HostDeleteVlanDeviceCmd{}
	ctx.GetCommand(cmd)

	DeleteVlanEth(cmd.VlanEthname)

	return nil
}

func GetAllBridgeInterface(bridgeName string) []string {
	bash := utils.Bash{
		Command: fmt.Sprintf("brctl show %s|sed -n '2,$p'|cut -f 6-10", bridgeName),
		NoLog: true,
	}
	_, out, _, _ := bash.RunWithReturn()
	vifs := strings.Split(out, "\n")
	for i, vif := range(vifs) {
		vifs[i] = strings.TrimSpace(vif)
	}
	return vifs
}

func hostDeleteBridgeHandler(ctx *server.CommandContext) interface{} {
	b := utils.NewBash()
	cmd := &HostDeleteBridgeCmd{}
	ctx.GetCommand(cmd)

	vifs := GetAllBridgeInterface(cmd.BridgeName)
	for _, vif := range(vifs) {
		b.Command = fmt.Sprintf("brctl delif %s %s", cmd.BridgeName, vif)
		b.Run()
	}

	b.Command = fmt.Sprintf("ip link set %s down", cmd.BridgeName)
	b.Run()

	b.Command = fmt.Sprintf("brctl delbr %s", cmd.BridgeName)
	b.Run()

	return nil
}


func GetDeviceIp(ethname string) string {
	bash := utils.Bash{
		Command: fmt.Sprintf("ip addr show dev %s|grep inet|grep -v inet6|awk -F'inet' '{print $2}'|awk '{print $1}'|awk -F'/24' '{print $1}'", ethname),
		NoLog: true,
	}
	_, out, _, err := bash.RunWithReturn()
	if err != nil {
		return ""
	} else {
		return out
	}
}

func hostSetDeviceIpHandler(ctx *server.CommandContext) interface{} {
	b := utils.NewBash()
	cmd := &HostSetDeviceIpCmd{}
	ctx.GetCommand(cmd)

	if isNetworkDeviceExisting(cmd.Ethname) {
		if GetDeviceIp(cmd.Ethname) != cmd.Ip {
			b.Command = fmt.Sprintf("ifconfig %s %s netmask %s", cmd.Ethname, cmd.Ip, cmd.Netmask)
			b.Run()
		}
	}

	return nil
}


func isNetworkDeviceExisting(ethname string) bool {
	bash := utils.Bash{
		Command: fmt.Sprintf("ip link show %s", ethname),
		NoLog: true,
	}
	returnCode, _, _, _ := bash.RunWithReturn()
	if returnCode == 0 {
		return true
	} else {
		return false
	}
}

func hostFlushDeviceIpHandler(ctx *server.CommandContext) interface{} {
	b := utils.NewBash()
	cmd := &HostFlushDeviceIpCmd{}
	ctx.GetCommand(cmd)

	if isNetworkDeviceExisting(cmd.Ethname) {
		b.Command = fmt.Sprintf("ip addr flush dev %s", cmd.Ethname)
		b.Run()
	}

	return nil
}


func hostShellCmdHandler(ctx *server.CommandContext) interface{} {
	cmd := &HostShellCmd{}
	ctx.GetCommand(cmd)
	var rsp HostShellRsp
	bash := utils.Bash{
		Command: cmd.Command,
		NoLog: true,
	}

	rsp.ReturnCode, rsp.Stdout, rsp.Stderr, _ = bash.RunWithReturn()
//	bash.PanicIfError()
	rsp.Command = cmd.Command
	rsp.Success = true

	return rsp
}

func hostVyosCmdHandler(ctx *server.CommandContext) interface{} {
	cmd := &HostShellCmd{}
	ctx.GetCommand(cmd)
	var rsp HostShellRsp

	rsp.Stdout = server.RunVyosScriptAsUserVyos(cmd.Command)
	rsp.Stderr = ""
	rsp.ReturnCode = 0
	rsp.Command = cmd.Command
	rsp.Success = true

	return rsp
}

func HostEntryPoint() {
	server.RegisterSyncCommandHandler(CREATE_VLAN_DEVICE_PATH, hostCreateVlanDeviceHandler)
	server.RegisterSyncCommandHandler(DELETE_VLAN_DEVICE_PATH, hostDeleteVlanDeviceHandler)
	server.RegisterSyncCommandHandler(DELETE_BRIDGE_PATH, hostDeleteBridgeHandler)
	server.RegisterSyncCommandHandler(SET_DEVICE_IP_PATH, hostSetDeviceIpHandler)
	server.RegisterSyncCommandHandler(FLUSH_DEVICE_IP_PATH, hostFlushDeviceIpHandler)
	server.RegisterSyncCommandHandler(HOST_SHELL_CMD_PATH, hostShellCmdHandler)
	server.RegisterSyncCommandHandler(HOST_VYOS_CMD_PATH, hostVyosCmdHandler)
//	server.RegisterSyncCommandHandler(HOST_ECHO_PATH, hostEchoHandler)
}
